session/types.proto

```diff
@@ -232,3 +233,140 @@ message ResponseVerificationHeader {
+message ObjectSessionContextV2 {
+  // Object request verbs.
+  enum Verb {
+    // Unknown verb.
+    VERB_UNSPECIFIED = 0;
+
+    // Refers to object.Put RPC call.
+    PUT = 1;
+
+    // Refers to object.Get RPC call.
+    GET = 2;
+
+    // Refers to object.Head RPC call.
+    HEAD = 3;
+
+    // Refers to object.Search RPC call.
+    SEARCH = 4;
+
+    // Refers to object.Delete RPC call.
+    DELETE = 5;
+
+    // Refers to object.GetRange RPC call.
+    RANGE = 6;
+
+    // Refers to object.GetRangeHash RPC call.
+    RANGEHASH = 7;
+  }
+
+  // Multiple verbs this context authorizes.
+  repeated Verb verbs = 1 [json_name = "verbs"];
+
+  // Target resource specification.
+  message Target {
+    // Container where operation is allowed.
+    neo.fs.v2.refs.ContainerID container = 1 [json_name = "container"];
+
+    // Specific objects where operation is allowed.
+    // Empty list means all objects in the container.
+    repeated neo.fs.v2.refs.ObjectID objects = 2 [json_name = "objects"];
+  }
+
+  // Target resource(s) this authorization applies to.
+  repeated Target targets = 2 [json_name = "targets"];
+}
+
+// ContainerSessionContextV2 carries context for ContainerService requests.
+// Extended version supporting multiple verbs.
+message ContainerSessionContextV2 {
+  // Container request verbs.
+  enum Verb {
+    // Unknown verb.
+    VERB_UNSPECIFIED = 0;
+
+    // Refers to container.Put RPC call.
+    PUT = 1;
+
+    // Refers to container.Delete RPC call.
+    DELETE = 2;
+
+    // Refers to container.SetExtendedACL RPC call.
+    SETEACL = 3;
+  }
+
+  // Multiple verbs this context authorizes.
+  repeated Verb verbs = 1 [json_name = "verbs"];
+
+  // If true, applies to all containers owned by the subject.
+  bool wildcard = 2 [json_name = "wildcard"];
+
+  // Particular container to which the action applies.
+  // Ignored if wildcard is true.
+  neo.fs.v2.refs.ContainerID container_id = 3 [json_name = "containerID"];
+}
+
+// SessionTokenV2 represents NeoFS Session Token with delegation support.
+message SessionTokenV2 {
+  // Session Token body.
+  message Body {
+    // Token version.
+    uint32 version = 1 [json_name = "version"];
+
+    // Token identifier (UUIDv4 in binary form).
+    bytes id = 2 [json_name = "id"];
+
+    // Account that issued this token (who signed it).
+    neo.fs.v2.refs.Account issuer = 3 [json_name = "issuer"];
+
+    // Accounts authorized by this token (who can use it).
+    repeated neo.fs.v2.refs.Account subjects = 4 [json_name = "subjects"];
+
+    // Lifetime of this token.
+    TokenLifetime lifetime = 5 [json_name = "lifetime"];
+
+    // Authorization context - what operations are allowed.
+    oneof context {
+      // ObjectService authorization context.
+      ObjectSessionContextV2 object = 6 [json_name = "object"];
+
+      // ContainerService authorization context.
+      ContainerSessionContextV2 container = 7 [json_name = "container"];
```

Comment:
Maybe we can have a single set of object/container operations? I'm not sure what this separation buys us.